modules/contact/test/tests/3d-mortar-contact/frictional-mortar-3d_pg.i

```diff
@@ -0,0 +1,535 @@
+    type = SideSetsAroundSubdomainGenerator
+    input = bottom_bottom_sideset
+    new_boundary = bottom_front
+    block = bottom_block
+    normal = '0 1 0'
+  []
+  [bottom_back_sideset]
+    type = SideSetsAroundSubdomainGenerator
+    input = bottom_front_sideset
+    new_boundary = bottom_back
+    block = bottom_block
+    normal = '0 -1 0'
+  []
+  [secondary]
+    input = bottom_back_sideset
+    type = LowerDBlockFromSidesetGenerator
+    sidesets = 'top_bottom' # top_back top_left'
+    new_block_id = '10001'
+    new_block_name = 'secondary_lower'
+  []
+  [primary]
+    input = secondary
+    type = LowerDBlockFromSidesetGenerator
+    sidesets = 'bottom_top'
+    new_block_id = '10000'
+    new_block_name = 'primary_lower'
+  []
+  uniform_refine = 0
+  allow_renumbering = false
+[]
+
+[Variables]
+  [mortar_normal_lm]
+    block = 'secondary_lower'
+    use_dual = true
+  []
+  [mortar_tangential_lm]
+    block = 'secondary_lower'
+    use_dual = true
+  []
+  [mortar_tangential_3d_lm]
+    block = 'secondary_lower'
+    use_dual = true
+  []
+[]
+
+[Modules/TensorMechanics/Master]
+  [all]
+    add_variables = true
+    strain = FINITE
+    block = '1 2'
+    use_automatic_differentiation = false
+    generate_output = 'stress_xx stress_xy stress_xz stress_yy stress_zz'
+  []
+[]
+
+[Materials]
+  [tensor]
+    type = ComputeIsotropicElasticityTensor
+    block = '1'
+    youngs_modulus = 1.0e4
+    poissons_ratio = 0.0
+  []
+  [stress]
+    type = ComputeFiniteStrainElasticStress
+    block = '1'
+  []
+
+  [tensor_1000]
+    type = ComputeIsotropicElasticityTensor
+    block = '2'
+    youngs_modulus = 1e5
+    poissons_ratio = 0.0
+  []
+  [stress_1000]
+    type = ComputeFiniteStrainElasticStress
+    block = '2'
+  []
+[]
+
+[UserObjects]
+  [weighted_vel_uo]
+    type = LMWeightedVelocitiesUserObject
+    primary_boundary = 'bottom_top'
+    secondary_boundary = 'top_bottom'
+    primary_subdomain = 'primary_lower'
+    secondary_subdomain = 'secondary_lower'
+    lm_variable_normal = mortar_normal_lm
+    lm_variable_tangential_one = mortar_tangential_lm
+    lm_variable_tangential_two = mortar_tangential_3d_lm
+    secondary_variable = disp_x
+    disp_x = disp_x
+    disp_y = disp_y
+    disp_z = disp_z
+    use_petrov_galerkin = true
+    aux_lm = aux_lm
+  []
+[]
+
+[Constraints]
+  [friction]
+    type = ComputeFrictionalForceLMMechanicalContact
+    primary_boundary = 'bottom_top'
+    secondary_boundary = 'top_bottom'
+    primary_subdomain = 'primary_lower'
+    secondary_subdomain = 'secondary_lower'
+    variable = mortar_normal_lm
+    disp_x = disp_x
+    disp_y = disp_y
+    disp_z = disp_z
+    use_displaced_mesh = true
+    mu = 0.4
+    c = 1e4
+    c_t = 1.0e4
+    friction_lm = mortar_tangential_lm
+    friction_lm_dir = mortar_tangential_3d_lm
+    weighted_gap_uo = weighted_vel_uo
+    weighted_velocities_uo = weighted_vel_uo
+    use_petrov_galerkin = true
+    aux_lm = aux_lm
```

Comment:
I haven't looked thoroughly yet into your code, but I wonder why we need to pass the PG settings to the constraint enforcement in the new user object design: Wouldn't the weighted quantities generated by the user object already have the desired PG result?

Reply:
IINW, the LM test functions are utilized in `WeightedGapUserObject` thus needing to be overridden when we want to use PG

Reply:
I agree with that statement. But you are also passing those to the constraint objects which are agnostic of the PG approach since they are just querying the user object for mortar quantities (which are indeed using PG).

Reply:
Oh I get your point. I saw `_test` initialized in the `MortarConstraintBase` so I just override it, not aware of it never been used later or in its inherited classes.

I'd propose we keep both or delete both in `MortarConstraintBase` - what do you think?

Reply:
Hmm I am not sure. If we keep it, it's not to be used within our contact mechanics approach. But maybe for other applications? 

@lindsayad please chime in, if it can potentially be useful for other applications within the constraint system we should probably keep it, otherwise remove it.

Reply:
Humm I saw that the `ComputeWeightedGapCartesianLMMechanicalContact` class is not converted to an UO yet. Maybe we should keep it in the `MortarConstraintBase` for now (and longer if there are other potential uses)?

Anyways I am removing excessive lines in the converted inputs

Reply:
Cool, thanks!!

Reply:
We must keep it. We use it for solution continuity and we still use it for thermal contact

Reply:
> We use it for solution continuity and we still use it for thermal contact

But for thermal contact and solution continuity the LM interpolation can already be of type Lagrange. Do you think those applications can also benefit from PG? (Not disagreeing at all with keeping it, just would like to understand that).

Reply:
I’m saying we need to keep the test member. I’d be a proponent for removing the petrov galerkin data from the constraint class until there is a user other than mechanical contact for PG

Reply:
Alternatively, I'd be curious about trying `use_dual` for the shape functions and the standard test functions for a solution continuity problem. Up to you

Reply:
Wait, I was confused.. We still need PG in MortarConstraintBase because the cartesian LM constraints need it.

Reply:
> 



> Alternatively, I'd be curious about trying `use_dual` for the shape functions and the standard test functions for a solution continuity problem. Up to you

The reason of having _test as standard basis is because we need to ensure positiveness of the weighted gap. On the other hand, keeping the dual basis for the shape function retains the diagonality of the mortar matrix thus easy variable condensation etc. Based on what I've learned from literature, we would loose both advantages if we do things the opposite.

Reply:
I was not proposing to do the opposite for solution continuity. In solution continuity, the Lagrange multiplier residuals do not depend on the Lagrange multiplier, similarly to mechanical contact when you are in contact.

In your PG method, `_test` or `test()` uses standard basis functions. And then the shapes that are underlying evaluation of `_lambda[_qp]` using the dual basis functions. Is my understanding correct? This is what I was proposing trying for solution continuity.

Reply:
I guess when using PG, the only place the dual shape functions appear is in evaluation of the contact pressure (the Lagrange multiplier) for application to the displacement residuals. Is that correct?

Reply:
> 



> 



> I was not proposing to do the opposite for solution continuity. In solution continuity, the Lagrange multiplier residuals do not depend on the Lagrange multiplier, similarly to mechanical contact when you are in contact.
> 
> In your PG method, `_test` or `test()` uses standard basis functions. And then the shapes that are underlying evaluation of `_lambda` uses the dual basis functions. Is my understanding correct? This is what I was proposing trying for solution continuity.

That's correct. Thanks for your clarification. Let me try solution continuity with it as well.

Reply:
> I guess when using PG, the only place the dual shape functions appear is in evaluation of the contact pressure (the Lagrange multiplier) for application to the displacement residuals. Is that correct?

From my understanding, yes.